### PR TITLE
MAINT: `linalg.logm`: avoid `ZeroDivisionError`, emit warning properly, avoid redundant squareness check

### LINKED ISSUE #17136: BUG: logm sometimes raises ZeroDivisionError for matrices with zero norm
### Describe your issue.

Running `scipy.linalg.logm` on a matrix with zero norm leads to inconsistent behaviour. For example, running
```python
logm(np.zeros((2, 2), dtype=int))
```
leads to the following result:
```
array([[-46.05170186,   0.        ],
       [  0.        , -46.05170186]])
```
with an appropriate warning  displayed ("logm result may be inaccurate, approximate err = inf"). 

However, the same matrix with floats leads to the ZeroDivisionError shown below. Note that this error actually occurs after the `logm` is calculated. The actual algorithm is consistent irrespective of data type. It happens while estimating the error of the algorithm (`errtest`).

I would suggest that the behaviour for the integer matrix is better. It would be easy to get that behaviour consistently across all zero-norm matrices: namely to check that `norm(A,1)` is zero before computing `errtest`. If `norm(A,1)` is zero `errtest` can be set to infinity (and the warning message will normally be displayed).

On a closely related note: I think returning an matrix of `NaN`s is a better behaviour for these zero-norm matrices (i.e., if  `errtest` is infinite) rather than the random numbers returned now.

### Reproducing Code Example

```python
from scipy.linalg import logm
import numpy as np
logm(np.zeros((2, 2), dtype=int))
logm(np.zeros((2, 2), dtype=float))
```

### Error message

```shell
ZeroDivisionError                         Traceback (most recent call last)
Cell In [7], line 1
----> 1 logm(np.zeros((2, 2), dtype=float))

File <python-dir>/python3.10/site-packages/scipy/linalg/_matfuncs.py:204, in logm(A, disp)
    202 errtol = 1000*eps
    203 #TODO use a better error approximation
--> 204 errest = norm(expm(F)-A,1) / norm(A,1)
    205 if disp:
    206     if not isfinite(errest) or errest >= errtol:

ZeroDivisionError: float division by zero
```

### SciPy/NumPy/Python version information

1.9.1 1.23.3 sys.version_info(major=3, minor=10, micro=6, releaselevel='final', serial=0)

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Closes gh-17136

#### What does this implement/fix?
gh-17136 noted a few issues in `linalg.logm`:

- If the norm of `A` is exactly zero and `lange` is used to evaluate it, the error estimate check raises a `ZeroDivisionError` because `lange` returns a Python `float` instead of a NumPy `float` of the appropriate dtype.
- There is a redundant shape check - it was being checked at the beginning of the function and immediately afterwards in `_logm`.

While working on this, I noticed that any warnings are `print`ed to the console rather than emitted via `warnings.warn`. 

This PR fixes these things.

#### Additional information
There are still issues here related to the an old-style API in which a `disp` parameter controls whether a warning is emitted or a tuple is returned... let's not get into that here. In the meantime, I am respecting that API: it only printed the warning when `disp=True`, so I only emit the warning when `disp=True`. Presumably emitting the warning with `warnings` rather than `print`ing is an acceptable change in behavior that can simply be documented in the release notes.

Comments in the issue suggest that if the error is high, the returned array should include NaNs rather than inaccurate values. It looks like this is because the function is not confident in its error estimate or tolerance. Since the result may still be useful to the user in some cases, I'm leaving it alone.

### CODE REVIEW

File: scipy/linalg/_matfuncs.py [hunk 1]
```diff
@@ -194,17 +195,19 @@ def logm(A, disp=True):
...
     errtol = 1000*eps
     # TODO use a better error approximation
-    errest = norm(expm(F)-A, 1) / norm(A, 1)
+    with np.errstate(divide='ignore', invalid='ignore'):
+        errest = norm(expm(F)-A, 1) / np.asarray(norm(A, 1), dtype=A.dtype).real[()]
     if disp:
         if not isfinite(errest) or errest >= errtol:
-            print("logm result may be inaccurate, approximate err =", errest)
+            message = f"logm result may be inaccurate, approximate err = {errest}"
+            warnings.warn(message, RuntimeWarning)
```
Comment (mdhaber) [inline scipy/linalg/_matfuncs.py:210]:
```suggestion
            warnings.warn(message, RuntimeWarning, stacklevel=2)
```

**Comment by mdhaber** [comment]
Previous CI run was successful except for lint failure; now fixed.

**Review by j-bowhay — APPROVED** [review]
Thanks @mdhaber, it seems unfortunate that the return type of `norm` is not consistent. But anyway this works as advertised!

**Comment by mdhaber** [comment]
Hmm that might be true, but if I remember correctly, the problem here was not inconsistency - the output type might always be a Python float. The trouble with a Python float here is that division by zero results in an error rather than `inf` or `nan`.
In any case, thanks for taking a look!

**Comment by j-bowhay** [comment]
I think my point was if this was the 2norm or the numpy implementation this wouldn't have been an issue
```
In [18]: a = np.zeros((2, 2), dtype=float)
In [19]: norm(a)
Out[19]: np.float64(0.0)

In [20]: norm(a, 1)
Out[20]: 0.0

In [21]: np.linalg.norm(a, 1)
Out[21]: np.float64(0.0)
```

**Comment by mdhaber** [comment]
Oops, I forgot. Yeah...

**Pull request merged**